### PR TITLE
[tosa] Legalize tfl.rfft2d to tosa.rfft2d

### PR DESCRIPTION
**Pull request by lhutton1**
In the case the provided fft_length > the input height or width, a tfl.pad operation is automatically inserted. When fft_length < the input height or width, a slice is manually inserted to crop the input.

~Note: this PR relies on commits from #59606 and therefore also contains the contents of #59606.~

Signed-off-by: Luke Hutton <luke.hutton@arm.com>

### CODE REVIEW

File: tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc [hunk 1]
```diff
@@ -4052,6 +4053,76 @@ LogicalResult ConvertTFLImagOp::matchAndRewrite(
...
+      rewriter.getDenseI64ArrayAttr(new_shape));
+  auto reshape_2 = CreateOpAndInfer<tosa::ReshapeOp>(
+      rewriter, loc, fp32_ty, rfft2d.getResult(1),
+      rewriter.getDenseI64ArrayAttr(new_shape));
+
+  llvm::SmallVector<Value, 2> values = {reshape_1, reshape_2};
+  auto concat = CreateOpAndInfer<tosa::ConcatOp>(rewriter, loc, fp32_ty, values,
+                                                 rewriter.getI64IntegerAttr(3));
+
+  Value cast = rewriter
```
Comment (jpienaar) [inline tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:4117]: replaceOpWithNewOp ?

**Comment by jpienaar** [comment]
(I don't know why AMD build failed here)

File: tensorflow/compiler/mlir/tosa/tests/lower-complex-types.mlir [hunk 2]
```diff
@@ -40,3 +40,26 @@ func.func @test_mixed_output(%arg0: tensor<1x4x4x2xf32>, %arg1: tensor<1x4x4xf32
...
+// CHECK-SAME: %[[VAL_0:.*]]: tensor<1x4x4x2xf32>
+// CHECK: %[[VAL_1:.*]] = "tosa.slice"(%[[VAL_0]]) {size = array<i64: 1, 4, 4, 1>, start = array<i64: 0, 0, 0, 0>} : (tensor<1x4x4x2xf32>) -> tensor<1x4x4x1xf32>
+// CHECK: %[[VAL_2:.*]] = "tosa.reshape"(%[[VAL_1]]) {new_shape = array<i64: 1, 4, 4>} : (tensor<1x4x4x1xf32>) -> tensor<1x4x4xf32>
+// CHECK: %[[VAL_3:.*]], %[[VAL_4:.*]] = "tosa.rfft2d"(%[[VAL_2]]) : (tensor<1x4x4xf32>) -> (tensor<1x4x3xf32>, tensor<1x4x3xf32>)
+// CHECK: %[[VAL_5:.*]] = "tosa.reshape"(%[[VAL_3]]) {new_shape = array<i64: 1, 4, 3, 1>} : (tensor<1x4x3xf32>) -> tensor<1x4x3x1xf32>
+// CHECK: %[[VAL_6:.*]] = "tosa.reshape"(%[[VAL_4]]) {new_shape = array<i64: 1, 4, 3, 1>} : (tensor<1x4x3xf32>) -> tensor<1x4x3x1xf32>
+// CHECK: %[[VAL_7:.*]] = "tosa.concat"(%[[VAL_5]], %[[VAL_6]]) {axis = 3 : i64} : (tensor<1x4x3x1xf32>, tensor<1x4x3x1xf32>) -> tensor<1x4x3x2xf32>
+// CHECK: return %[[VAL_7]] : tensor<1x4x3x2xf32>
+func.func @test_complex_input_output_op(%arg0: tensor<1x4x4xcomplex<f32>>) -> (tensor<1x4x3xcomplex<f32>>) {
+    %0 = builtin.unrealized_conversion_cast %arg0 : tensor<1x4x4xcomplex<f32>> to tensor<1x4x4x2xf32>
```
Comment (jpienaar) [inline tensorflow/compiler/mlir/tosa/tests/lower-complex-types.mlir:56]: It's a little bit weird to see an unrealized conversion cast here, normally they only occur as part of dialect conversion and not before or after the conversion. Is there a way for the user to perform this transform?

**Review by jpienaar — COMMENTED** [review]
Overall looks good

**Review by lhutton1 — COMMENTED** [review]
Thanks for the review @jpienaar, the AMD failure looked unrelated hopefully it passes this time

**Comment by lhutton1** [comment]
friendly ping if you have some free time :) - the AMD build failure doesn't look related

**Review by jpienaar — APPROVED** [review]
LG and Ilij general thanks!

File: tensorflow/compiler/mlir/tosa/tests/lower-complex-types.mlir (hunk 2 above)
Comment (jpienaar) [reply tensorflow/compiler/mlir/tosa/tests/lower-complex-types.mlir:56]: SGTM.

**Review by jpienaar — APPROVED** [review]

**Review by jpienaar — APPROVED** [review]

**Pull request merged**